tools/ping/src/lib.rs

```diff
@@ -147,6 +156,10 @@ impl AppInfo {
         let timestamp = time::Instant::now();
         let timeout = timestamp + PING_TIMEOUT;
 
+        let account_id = self.peer_id_to_account_id(&peer_id);
+        crate::metrics::PING_SENT.with_label_values(&[&peer_str(peer_id, account_id)]).inc();
+        self.push_metrics();
```

Comment:
so here push_metrics() is a blocking call that actually does take a nontrivial amount of time. that feels like kind of a shame because so far there's only one thread doing stuff, and if it's blocked on this call while some `Pong` message comes, then the latencies will be inflated. since this is probably running on a machine w more than one CPU why not just spawn another task and push metrics every 5 seconds or something? like:

```diff
diff --git a/tools/ping/src/lib.rs b/tools/ping/src/lib.rs
index 22785aa31..1979e7f80 100644
--- a/tools/ping/src/lib.rs
+++ b/tools/ping/src/lib.rs
@@ -123,23 +123,15 @@ struct AppInfo {
     requests: BTreeMap<PingTarget, HashMap<Nonce, PingTimes>>,
     timeouts: BTreeSet<PingTimeout>,
     account_filter: Option<HashSet<AccountId>>,
-    metrics_gateway_address: Option<String>,
-    chain_id: String,
 }
 
 impl AppInfo {
-    fn new(
-        chain_id: &str,
-        account_filter: Option<HashSet<AccountId>>,
-        metrics_gateway_address: Option<String>,
-    ) -> Self {
+    fn new(account_filter: Option<HashSet<AccountId>>) -> Self {
         Self {
             stats: HashMap::new(),
             requests: BTreeMap::new(),
             timeouts: BTreeSet::new(),
             account_filter,
-            metrics_gateway_address,
-            chain_id: chain_id.to_string(),
         }
     }
 
@@ -158,7 +150,6 @@ impl AppInfo {
 
         let account_id = self.peer_id_to_account_id(&peer_id);
         crate::metrics::PING_SENT.with_label_values(&[&peer_str(peer_id, account_id)]).inc();
-        self.push_metrics();
 
         match self.stats.entry(peer_id.clone()) {
             Entry::Occupied(mut e) => {
@@ -325,23 +316,21 @@ impl AppInfo {
     fn peer_id_to_account_id(&self, peer_id: &PeerId) -> Option<&AccountId> {
         self.stats.get(peer_id).and_then(|s| s.account_id.as_ref())
     }
+}
 
-    fn push_metrics(&self) {
-        let metric_families = prometheus::gather();
-        if let Some(metrics_gateway_address) = &self.metrics_gateway_address {
-            prometheus::push_metrics(
-                "near_ping",
-                prometheus::labels! {
-                    "chain_id".to_owned() => self.chain_id.clone(),
-                },
-                &metrics_gateway_address,
-                metric_families,
-                None,
-            )
-            .unwrap_or_else(|err| tracing::error!("Failed to push metrics: {:#?}", err));
-            tracing::debug!("Pushed metrics to {}", metrics_gateway_address);
-        }
-    }
+fn push_metrics(metrics_gateway_address: &str, chain_id: &str) {
+ ...
+ ...
 }
 
 fn handle_message(
@@ -415,7 +404,7 @@ async fn ping_via_node(
     ping_stats: &mut Vec<(PeerIdentifier, PingStats)>,
     metrics_gateway_address: Option<String>,
 ) -> anyhow::Result<()> {
-    let mut app_info = AppInfo::new(chain_id, account_filter, metrics_gateway_address);
+    let mut app_info = AppInfo::new(account_filter);
 
     app_info.add_peer(&peer_id, None);
 
@@ -452,6 +441,15 @@ async fn ping_via_node(
         }
     };
 
+    let chain_id = chain_id.to_string();
+    let _push_metrics = metrics_gateway_address.map(|url| {
+        tokio::spawn(async move {
+            loop {
+                push_metrics(&url, &chain_id);
+                tokio::time::sleep(std::time::Duration::from_secs(5)).await;
+            }
+        })
+    });
     let mut result = Ok(());
     let mut nonce = 1;
     let next_ping = tokio::time::sleep(std::time::Duration::ZERO);
```